### ISSUE TITLE
A question about the loss

### ISSUE THREAD

**Issue by Resurgamm** [yanweiyue/GDesigner#1]
From the code `single_loss = - log_prob * utility`, we know that if the answer is wrong, then the single_loss = 0, and if the answer is correct, then single_loss > 0 since - log_prob > 0.

Why is this correct? I mean, shouldn't the loss be bigger for wrong answers?

**Comment by yanweiyue** [comment]
Thank you very much for your attention and detailed research on our work.

In policy gradient methods, the goal is to optimize a policy by adjusting its parameters to maximize the expected cumulative reward. To do this, we typically rely on sampling techniques to estimate the expected reward because directly computing this expectation is often intractable. The objective of policy gradient methods is to maximize the **expected reward**:
$`\nabla_\theta J(\theta) \approx \frac{1}{N} \sum_{i=1}^N \nabla_\theta \log \pi_\theta(a_i) R_{i}`$

In our code `loss = mean(single_loss) = mean(- log_prob * utility) = - mean(log_prob * utility)`, the term `mean(log_prob * utility)` serves as an estimate of the **expected reward**. By minimizing the loss, we effectively maximize the **expected reward**.

I hope my response has clarified your doubts. If you have any further questions, please feel free to ask.

**Comment by Resurgamm** [comment]
Thanks for your brief explanation. I have already known and agreed with the statement that by minimizing the loss, we effectively maximize the expected reward. However, I still have a confusion that is, like I asked before, since log_prob < 0, a correct answer will lead to a negative value of the expected reward, and thus a larger positive loss. 

Another question is that when I run the training code, The loss curve appears to be oscillating. To solve this problem, I printed out the parameters of the gcn, and I found that the gradient value of the second layer is too small that it's a 1e-12 weight class even if it's the first round. Is it because of this that the model cannot be optimized and the loss curve oscillates, or there exists som other reasons? And how can I solve this question so that the model can be well optimized?

I will be appreciated if you can answer my questions.

**Comment by Resurgamm** [comment]
Thanks for your brief explanation. I have already known and agreed with the statement that by minimizing the loss, we effectively maximize the expected reward. However, I still have a confusion that is, like I asked before, since log_prob < 0, a correct answer will lead to a negative value of the expected reward, and thus a larger positive loss. 

Another question is that when I run the training code, The loss curve appears to be oscillating. To solve this problem, I printed out the parameters of the gcn, and I found that the gradient value of the second layer is too small that it's a 1e-12 weight class even if it's the first round. Is it because of this that the model cannot be optimized and the loss curve oscillates, or there exists som other reasons? And how can I solve this question so that the model can be well optimized?

I will be appreciated if you can answer my questions.

**Comment by Resurgamm** [comment]
I'm sorry about the gradient disappearing, as I said before. In fact, by magnifying loss with the code `loss = loss * 100`, I found that the gradient value was only relatively small and did not disappear. However, the loss curve still oscillates after amplification, which means that the training still has no effect. 

I would like to know if you encountered this problem while doing the experiment, and if so, how you solved it and why it occurred.

Your explanation will be useful for me since I'm eager to replicate the experiment and successfully train the model. I would appreciate it if you could answer my questions.

I ran GDesigner on MMLU by running the following scripts with the model `gpt-4o-mini`.

```bash
python experiments/run_mmlu.py --mode FullConnected --batch_size 10 --agent_nums 6 --num_iterations 100 --num_rounds 1 --optimized_spatial
```

I've attached the loss curve below. 

![Image](https://github.com/user-attachments/assets/1fb63299-6f59-4080-b93a-be0c9d4b7d1c)

**Comment by yanweiyue** [comment]
Apologies for the delayed reply. We greatly appreciate your thorough research and continued attention to our work.

Regarding your comment that `"a correct answer will lead to a negative value of the expected reward, and thus a larger positive loss"`, in the context of the policy gradient algorithm used in this paper, the process of deriving the reward from the answer is not learnable. You can think of the reward for an answer as a fixed value. The parameters that the Policy Network can optimize during gradient descent only affect the probability of the answer, and our learning objective is to maximize the probability of high-reward answers.

As for the `"loss curve still oscillating after amplification,"` we acknowledge that this phenomenon does indeed occur, as the loss in reinforcement learning is closely related to the queries faced by the agent. A better system can only guarantee that the loss will be relatively smaller for the same queries. I'm not entirely sure whether the x-axis in the figure you provided represents full epochs or different batches of queries. If the x-axis represents different queries, we consider the oscillating loss to be acceptable.

To demonstrate that our optimization is indeed effective, you can refer to `"5.4 Robustness Verification."` When faced with misleading agent attacks, our algorithm can effectively identify them and ensure accuracy.

Lastly, we sincerely thank you for your insightful questions. We will continue to improve our algorithm and reinforce our training methods based on your feedback. Your suggestions have been incredibly helpful in refining this paper. 

Once again, we apologize for the delayed reply and appreciate your valuable insights.

**Comment by Resurgamm** [comment]
In the figure, the x-axis represents one training round with a batch size of 10. For the `I'm not entirely sure whether the x-axis in the figure you provided represents full epochs or different batches of queries.`, my understanding is that it seems there should be `different batches of queries.` So, according to your interpretation, this is normal, right?

Thanks again for your explanation, it helps me a lot.

**Comment by Resurgamm** [comment]
And another question is about the ACC. I ran GDesigner on MMLU by running the following scripts with the model `gpt-3.5-turbo`. 

```bash
python experiments/run_mmlu.py --mode FullConnected --batch_size 4 --agent_nums 6 --num_iterations 1000 --num_rounds 1 --optimized_spatial
```

I've attached the curves below. The model is evaluated every ten rounds of training.

![Image](https://github.com/user-attachments/assets/8e2a2b24-6590-487b-9cf4-16278d1194d4)

So, according to your explanation, we consider the oscillating loss to be acceptable. However, the ACC curve is around 0.6 instead of 0.8515 mentioned in the paper, and with the training process, the ACC didn't increase. We think that it had an ACC of about 0.6 because even without training, we still have end-to-end language models and a predefined communication structure. 

So why hasn't ACC increased with training?

**Comment by yanweiyue** [comment]
Regarding the accuracy issue you mentioned, in Table 1 of our paper, the accuracy of G-designer on the MMLU dataset is reported as 84%, based on GPT-4. It seems that we did not report any experimental results based on GPT-3.5 in the paper. 

It's worth noting that the improvement from multi-agent collaboration seems less significant on knowledge-based datasets like MMLU, and the performance is largely dependent on the knowledge and capabilities inherent in the base LLM.

I apologize for any confusion this may have caused. We will continue to study and optimize our training strategies moving forward. Thank you for your insightful suggestion!
